File: test/cpp-raw/stateful.cpp

#include "../cpp-utility/os.hpp"
#include "../dependencies/catch.hpp"

#include "testlib.hpp"

TEST_CASE( "stateful" )
{
    // Initialize COM.
    InitializeRuntime();

    // Get the IResultOperations interface.
    IStatefulOperations_Automation* pOps = nullptr;
    intercom::HRESULT hr = CreateInstance(
            CLSID_StatefulOperations,
            IID_IStatefulOperations_Automation,
            &pOps );

    REQUIRE( hr == intercom::SC_OK );
    REQUIRE( pOps != nullptr );

    SECTION( "State is stored" )
    {
        pOps->PutValue( 10 );
        REQUIRE( pOps->GetValue() == 10 );
        pOps->PutValue( -100 );
        REQUIRE( pOps->GetValue() == -100 );
        pOps->PutValue( 55555 );
        REQUIRE( pOps->GetValue() == 55555 );
    }

    REQUIRE( pOps->Release() == 0 );

    UninitializeRuntime();
}

File: test/testlib/src/stateful.rs
use intercom::*;

#[com_class(StatefulOperations)]
pub struct StatefulOperations
{
    state: i32,
}

impl Default for StatefulOperations {
    fn default() -> Self {
        StatefulOperations { state: 0xABBACD }
    }
}

#[com_interface]
#[com_impl]
impl StatefulOperations
{
    pub fn put_value(&mut self, v: i32)
    {
        self.state = v;
    }
    pub fn get_value(&mut self) -> i32
    {
        self.state
    }
}
